core/src/main/java/com/linecorp/armeria/internal/common/stream/HttpDataDeframer.java

```diff
@@ -0,0 +1,417 @@
+     *
+     * <p>Note that The {@link HttpData}s which was fully consumed will be cleaned up automatically
+     * so that the {@link Queue} does not grow up infinitely.
+     * The {@code readerIndex} could be decreased after the cleanup process.
+     */
+    protected abstract List<T> process(HttpDataReader reader);
+
+    @Override
+    public final void onSubscribe(Subscription subscription) {
+        requireNonNull(subscription, "subscription");
+        if (upstreamUpdater.compareAndSet(this, null, subscription)) {
+            deferredInit();
+        } else {
+            subscription.cancel();
+        }
+    }
+
+    @Override
+    public final void subscribe(Subscriber<? super T> subscriber) {
+        super.subscribe(subscriber);
+        if (subscribedUpdater.compareAndSet(this, 0, 1)) {
+            deferredInit();
+        }
+    }
+
+    private void deferredInit() {
+        if (upstream != null && subscribed != 0) {
+            if (initializedUpdater.compareAndSet(this, 0, 1)) {
+                upstream.request(1);
+            }
+        }
+    }
+
+    @Override
+    public final void onNext(HttpData data) {
+        if (eventLoop.inEventLoop()) {
+            onNext0(data);
+        } else {
+            eventLoop.execute(() -> onNext0(data));
+        }
+    }
+
+    private void onNext0(HttpData data) {
+        try {
+            queue.add(data.byteBuf());
+            final List<T> decoded = process(reader);
+
+            if (decoded == null || decoded.isEmpty()) {
+                upstream.request(1);
+            } else {
+                for (T item : decoded) {
+                    write(item);
+                }
+                whenConsumed().thenRun(() -> upstream.request(1));
+            }
+            reader.discardReadBytes();
+        } catch (Throwable ex) {
+            Exceptions.throwIfFatal(ex);
+            cancelAndCleanup();
+            abort(ex);
+        }
+    }
+
+    private void cancelAndCleanup() {
+        isCancelled = true;
+        upstream.cancel();
+        cleanup();
+    }
+
+    @Override
+    public final void onError(Throwable cause) {
+        requireNonNull(cause, "cause");
+        if (isCancelled) {
+            return;
+        }
+        cleanup();
+        abort(cause);
+    }
+
+    @Override
+    public final void onComplete() {
+        if (isCancelled) {
+            return;
+        }
+        cleanup();
+        close();
+    }
+
+    private void cleanup() {
+        if (!eventLoop.inEventLoop()) {
+            eventLoop.execute(this::cleanup);
+            return;
+        }
+
+        for (ByteBuf buf : queue) {
+            buf.release();
+        }
+        queue.clear();
+    }
+
+    /**
+     * A reader of a stream of {@link HttpData} which was delivered from the {@link Publisher}.
+     * This class is NOT thread safe.
+     */
+    protected final class HttpDataReader {
```

Comment:
- Perhaps better declaring a top-level reader interface and keeping a private implementation here?
- How about `HttpDeframerInput`?